### PR TITLE
Support REMOVE a:L in cost planner

### PR DESCRIPTION
**Pull request by pontusmelke**

### CODE REVIEW

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/ast/convert/plannerQuery/ClauseConverters.scala [hunk 1]
```diff
@@ -312,4 +313,15 @@ object ClauseConverters {
...
   }
+
+    def addRemoveToLogicalPlanInput(acc: PlannerQueryBuilder, clause: Remove): PlannerQueryBuilder = {
+      clause.items.foldLeft(acc) {
+        // REMOVE n:Foo
+        case (builder, RemoveLabelItem(identifier, labelNames)) =>
+          builder.amendUpdateGraph(ug => ug.addRemoveLabelPatterns(RemoveLabelPattern(IdName.fromIdentifier(identifier), labelNames)))
+
+        case (builder, _) =>
+          throw new CantHandleQueryException("Remove supported yet")
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/ast/convert/plannerQuery/ClauseConverters.scala]: The error message is a bit unhelpful here.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RemoveLabelsPipe.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  override def symbols = src.symbols
+
+  override def withEstimatedCardinality(estimated: Double) = copy()(Some(estimated))
+
+  override def dup(sources: List[Pipe]): Pipe = {
+    val (onlySource :: Nil) = sources
+    SetLabelsPipe(onlySource, identifier, labels)(estimatedCardinality)
+  }
+
+  override def localEffects = Effects(labels.map(label => SetLabel(label.name)): _*)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RemoveLabelsPipe.scala:61]: This is wrong, but then again it isn't used, so it doesn't matter that much. Perhaps we should just stop overriding `localEffects` for the new pipes we create?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RemoveLabelsPipe.scala [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  override def planDescriptionWithoutCardinality = src.planDescription.andThen(this.id, "RemoveLabels", identifiers)
+
+  override def symbols = src.symbols
+
+  override def withEstimatedCardinality(estimated: Double) = copy()(Some(estimated))
+
+  override def dup(sources: List[Pipe]): Pipe = {
+    val (onlySource :: Nil) = sources
+    SetLabelsPipe(onlySource, identifier, labels)(estimatedCardinality)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RemoveLabelsPipe.scala]: Busted for copy-pasting!

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RepeatableReadPipe.scala [hunk 4]
```diff
@@ -35,7 +35,8 @@ case class RepeatableReadPipe(src: Pipe)(val estimatedCardinality: Option[Double
   protected def internalCreateResults(input: Iterator[ExecutionContext], state: QueryState): Iterator[ExecutionContext] = {
     val cached = state.repeatableReads.getOrElseUpdate(this, input.toList)
 
-    cached.toIterator
+    //cached results must propagate initial results
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RepeatableReadPipe.scala:38]: I don't understand what this comment means. Is is possible to clarify it a bit?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/PlannerQuery.scala [hunk 5]
```diff
@@ -60,7 +60,9 @@ case class PlannerQuery(queryGraph: QueryGraph = QueryGraph.empty,
 
   def writes: Boolean = exists(_.updateGraph.nonEmpty)
 
-  def writesNodes: Boolean = exists(_.updateGraph.nodePatterns.nonEmpty)
+  def createNodes: Boolean = exists(_.updateGraph.nodePatterns.nonEmpty)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/PlannerQuery.scala]: create`s`Nodes

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala [hunk 6]
```diff
@@ -50,21 +50,46 @@ case class UpdateGraph(mutatingPatterns: Seq[MutatingPattern] = Seq.empty) {
...
+
   def patternNodeLabels: Map[IdName, Set[LabelName]] =
     nodePatterns.map(p => p.nodeName -> p.labels.toSet).toMap
 
-  def labels: Set[LabelName] = nodePatterns.flatMap(_.labels).toSet
+  def writeLabels: Set[LabelName] = nodePatterns.flatMap(_.labels).toSet
+
+  def removeLabels: Set[LabelName] = removeLabelPatterns.flatMap(_.labels).toSet
+
+  def removeLabelsOnNode(idName: IdName): Set[LabelName] = removeLabelPatterns.collect {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala]: I suggest `labelsToRemoveForNode`

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala [hunk 7]
```diff
@@ -50,21 +50,46 @@ case class UpdateGraph(mutatingPatterns: Seq[MutatingPattern] = Seq.empty) {
...
+    case p: RemoveLabelPattern => p
+  }
+
   def patternNodeLabels: Map[IdName, Set[LabelName]] =
     nodePatterns.map(p => p.nodeName -> p.labels.toSet).toMap
 
-  def labels: Set[LabelName] = nodePatterns.flatMap(_.labels).toSet
+  def writeLabels: Set[LabelName] = nodePatterns.flatMap(_.labels).toSet
+
+  def removeLabels: Set[LabelName] = removeLabelPatterns.flatMap(_.labels).toSet
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala]: I suggest `labelsToRemove`

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala [hunk 8]
```diff
@@ -50,21 +50,46 @@ case class UpdateGraph(mutatingPatterns: Seq[MutatingPattern] = Seq.empty) {
...
+
+  def removeLabelsOnNode(idName: IdName): Set[LabelName] = removeLabelPatterns.collect {
+    case RemoveLabelPattern(n, labels) if n == idName => labels
+  }.flatten.toSet
 
   def relTypes: Set[RelTypeName] = relPatterns.map(_.relType).toSet
 
+  def updateNodes = nodePatterns.nonEmpty || removeLabelPatterns.nonEmpty
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala]: I suggest update`s`Nodes

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala [hunk 9]
```diff
@@ -50,21 +50,46 @@ case class UpdateGraph(mutatingPatterns: Seq[MutatingPattern] = Seq.empty) {
...
+      (qg.patternNodeLabels.values.flatten.toSet intersect writeLabels).nonEmpty // CREATE(:A:B) MATCH(:B:C)?
+  }
+
+  private def removeLabelOverlap(qg: QueryGraph) = {
+    removeLabelPatterns.exists {
+      case RemoveLabelPattern(removeId, labelsToRemove) =>
+        //does any other identifier match on the labels I am deleting?
+        //MATCH (a:BAR)..(b) REMOVE b:BAR
+        labelsToRemove.exists(l => {
+          val otherLabelsRead = qg.patternNodes.filterNot(_ == removeId).flatMap(qg.knownLabelsOnNode)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/UpdateGraph.scala]: Is there a specific reason for using `knownLabelsOnNode` and not `allKnownLabelsOnNode` here?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanSingleQuery.scala [hunk 10]
```diff
@@ -62,18 +62,17 @@ case class PlanSingleQuery(planPart: (PlannerQuery, LogicalPlanningContext, Opti
       val leaves = plan.leaves.collect {
         case n: NodeLogicalLeafPlan => n.idName
       }
-
-
       //if we have unsafe rels we need to check relation overlap and delete
       //overlap immediately
       (hasUnsafeRelationships(plannerQuery) &&
-        (relationshipOverlap(plannerQuery) || deleteOverlap(plannerQuery))) ||
+        (plannerQuery.updateGraph.relationshipOverlap(plannerQuery.queryGraph) ||
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanSingleQuery.scala:68]: I changed my mind on those, but apparently you didn't :) I'm fine with this though.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanSingleQuery.scala [hunk 11]
```diff
@@ -82,13 +81,22 @@ case class PlanSingleQuery(planPart: (PlannerQuery, LogicalPlanningContext, Opti
    * with the labels updated in this query. This may cause the read to affected
    * by the writes.
    */
-  private def nodeOverlap(start: IdName, plannerQuery: PlannerQuery): Boolean = {
+  private def nodeOverlap(start: IdName, plannerQuery: PlannerQuery, stableId: IdName): Boolean = {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanSingleQuery.scala:84]: I like this concept of a stable id!

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 12]
```diff
@@ -965,6 +965,42 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
...
+    assertStats(result, labelsAdded = 2, nodesCreated = 4)
+    assertNumberOfEagerness(query, 1)
+  }
+
+  test("setting label in tail should be eager if overlap within tail") {
+    createNode()
+    createNode()
+    createLabeledNode("Foo")
+    createLabeledNode("Foo")
+    val query = "MATCH (n) CREATE (m) WITH * MATCH (o) SET o:Foo RETURN count(*)"
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala]: I don't see the overlap in this query?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 13]
```diff
@@ -1155,6 +1191,101 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
...
+    createLabeledNode("Bar")
+    createLabeledNode("Bar")
+    val query = "MATCH (n) CREATE (m) WITH * MATCH (o:Bar) REMOVE n:Foo RETURN count(*)"
+
+    val result = updateWithBothPlanners(query)
+    assertStats(result, labelsRemoved = 2, nodesCreated = 4)
+    assertNumberOfEagerness(query, 0)
+  }
+
+  test("undirected expand followed by remove label need not to be eager") {
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala]: what about `needn't be eager` instead?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RemoveLabelsPipe.scala (hunk 2 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RemoveLabelsPipe.scala:61]: Yeah but this is at least the same effects as `LabelAction` so it is at least equally wrong!

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RemoveLabelsPipe.scala (hunk 3 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/RemoveLabelsPipe.scala]: oh the shame!

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanSingleQuery.scala (hunk 10 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanSingleQuery.scala:68]: yeah I don't have a strong opinion. We'll go back and forth until something sticks

**Comment by Mats-SX** [comment]
Build failure is unrelated. I've restarted the build.

**Pull request merged**